rust/lance/src/dataset/optimize.rs

```diff
@@ -784,7 +784,10 @@ async fn rewrite_files(
     // information.
     let fragments = migrate_fragments(dataset.as_ref(), &task.fragments, recompute_stats).await?;
     let mut scanner = dataset.scan();
-    scanner.with_fragments(fragments.clone()).with_row_id();
+    scanner
+        .with_fragments(fragments.clone())
+        .scan_in_order(true)
```

Comment:
I'm not sure this impacts the bug, but I should point out that `scan_in_order` guarantees that results will be in order of local row ids, but not in order of fragment ids. This is because the fragment list in the manifest is not guaranteed to be sorted by row_id. You should be able to force a scan to be globally sorted by row_id by passing in the `.fragments()` parameter with the list of fragments sorted by fragment id.